sdk/cosmos/azure-cosmos-encryption/src/main/java/com/azure/cosmos/EncryptionCosmosAsyncContainer.java

```diff
@@ -0,0 +1,179 @@
+import com.azure.cosmos.implementation.Utils;
+import com.azure.cosmos.implementation.encryption.CosmosResponseFactoryCore;
+import com.azure.cosmos.implementation.encryption.DecryptionResult;
+import com.azure.cosmos.implementation.encryption.EncryptionItemRequestOptions;
+import com.azure.cosmos.implementation.encryption.Encryptor;
+import com.azure.cosmos.implementation.guava25.base.Preconditions;
+import com.azure.cosmos.models.CosmosItemRequestOptions;
+import com.azure.cosmos.models.CosmosItemResponse;
+import com.azure.cosmos.models.EncryptionBridgeInternal;
+import com.azure.cosmos.models.PartitionKey;
+import reactor.core.publisher.Mono;
+import reactor.core.scheduler.Schedulers;
+
+import java.util.function.Consumer;
+
+// TODO: for now basic functionality is in. some APIs and some logic branch is not complete yet.
+public class EncryptionCosmosAsyncContainer extends CosmosAsyncContainer {
+    private final Encryptor encryptor;
+    private final CosmosResponseFactory resposeFactory = new CosmosResponseFactoryCore();
+
+    EncryptionCosmosAsyncContainer(String id, CosmosAsyncDatabase database, Encryptor encryptor) {
+        super(id, database);
+        this.encryptor = encryptor;
+    }
+
+    private Mono<CosmosItemResponse<byte[]>> createItemStream(byte[] payload,
+                                                              PartitionKey partitionKey,
+                                                              CosmosItemRequestOptions requestOptions) {
+        Preconditions.checkNotNull(payload, "payload can't be null");
+
+        // TODO: add diagnostics
+        EncryptionItemRequestOptions encryptionItemRequestOptions = Utils.as(requestOptions, EncryptionItemRequestOptions.class);
+        if (encryptionItemRequestOptions != null && encryptionItemRequestOptions.getEncryptionOptions() != null) {
+
+            payload = EncryptionProcessor.encryptAsync(payload, encryptor, encryptionItemRequestOptions.getEncryptionOptions());
+
+            Mono<CosmosItemResponse<byte[]>> response = super.createItem(payload, partitionKey, requestOptions);
+
+            return response
+                .subscribeOn(Schedulers.elastic())
+                .publishOn(Schedulers.elastic())
+                .map(rsp -> {
+                    EncryptionBridgeInternal.setByteArrayContent(rsp, decryptResponseAsync(EncryptionBridgeInternal.getByteArrayContent(rsp), encryptionItemRequestOptions.getDecryptionResultHandler()));
+                    return rsp;
+                }
+            );
+
+        } else {
+            return null;
+            // TODO
+            // return super.createItem()
+        }
+    }
+
+    // TODO ensure all other apis call this guy
+    public <T> Mono<CosmosItemResponse<T>> createItem(T item,
+                                                      PartitionKey partitionKey,
+                                                      CosmosItemRequestOptions option) {
```

Comment:
minor, `option` ->  `requestOptions`  to be more consistent with other APIs.

Reply:
thanks.